azure-core/azure/core/pipeline_client_async.py

```diff
@@ -0,0 +1,213 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+
+import logging
+import os
+import sys
+try:
+    from urlparse import urljoin, urlparse
+except ImportError:
+    from urllib.parse import urljoin, urlparse
+import warnings
+import xml.etree.ElementTree as ET
+
+from typing import List, Any, Dict, Union, IO, Tuple, Optional, Callable, Iterator, cast, TYPE_CHECKING  # pylint: disable=unused-import
+
+from .pipeline import AsyncPipeline
+from .pipeline.policies import ContentDecodePolicy
+from .pipeline.transport import HttpRequest, AioHttpTransport
+
+_LOGGER = logging.getLogger(__name__)
+
+class AsyncPipelineClient(object):
+    """Service client core methods.
+
+    This contains methods are sans I/O and not tight to sync or async implementation.
+    :param Configuration config: Service configuration.
+    """
+
+    def __init__(self, base_url, config, **kwargs):
+        if config is None:
+            raise ValueError("Config is a required parameter")
+        self._config = config
+        self._base_url = base_url
+        self._transport = kwargs.get('transport', AioHttpTransport(config))
+        self._pipeline = kwargs.get('pipeline', self._build_pipeline(config))
```

Comment:
This will call `_build_pipeline` even when `kwargs['pipeline']` has a value. (And, for some reason obscure to me, `self._pipeline` is assigned `None` in that case 🤔)

Similarly for the above assignment to `_transport`. That one's further troublesome because if the caller passed in a pipeline, that pipeline should already have a transport, and we should expect `kwargs['transport'] is None`. Also, `self._transport` is only used in `_build_pipeline`, maybe it should just be a parameter of that method?

Reply:
Good call, so I'm thinking it could be revised to something like:
```python
if kwargs.get('pipeline'):
    self._pipeline = kwargs['pipeline']
else:
    try:
        transport = kwargs['transport']
    except KeyError:
        transport = AioHttpTransport(config)
    self._pipeline = self._build_pipeline(config, transport)
```
So we no longer keep transport as an attribute on self, and instead pass it in to build the pipeline. What do you think?

Reply:
Looks good to me.